### ISSUE TITLE
Reusable Global Helper Functions

### ISSUE THREAD

**Issue by artembeep** [ajv-validator/ajv#1959]
**What version of Ajv you are you using?** 
Latest

**What problem do you want to solve?** 
Decrease compiled validator code size

**What do you think is the correct solution to problem?**
With definition of a custom `CodeKeywordDefinition` is it possible using `gen` scope to define a reusable function that would be declared once in a global scope of a schema validator compilation instead of inline reinjection the same declaration clone per each keyword reference.

E.g. for a schema
```js
{
    id: 'someSchema',
    properties:  {
         a: {customKeyword: 'options'},
         b: {customKeyword: 'options'},
     }
}
```
instead of compilation (pseudo code)
```js
const  someSchemaValidator = (data, ...) => {
   // a declaration
        function customKeywordHelper0() {.......}
        if (customKeywordHelper0(data.a)) {
            errors = [{path: '/a', keyword: 'customKeyword'}]
        }
   // b declaration
        function customKeywordHelper1() {.......}
        if (customKeywordHelper1(data.b)) {
            errors = [{path: '/b', keyword: 'customKeyword'}]
        }
}
```
It would be nice to have
```js
const  someSchemaValidator = (data, ...) => {
  // global helper reusable declaration 
   function customKeywordHelper() {.......}

   // a declaration
        if (customKeywordHelper(data.a)) {
            errors = [{path: '/a', keyword: 'customKeyword'}]
        }
   // b declaration
        if (customKeywordHelper(data.b)) {
            errors = [{path: '/b', keyword: 'customKeyword'}]
        }
}
```

**Will you be able to implement it?** 
I'm curious whether it is feasible in terms of current implementation. Maybe there's something `it.root.gen` available for this purpose?
If not I can assist after a high level  introduction into the most appropriate way to implement this since I'm not really familiar with the internal design for now .

P.S. Do I see it correct  as `CodeKeywordDefinition` is something the former `inline` keywords definition are evolved with time to?

Thank you.

**Comment by epoberezkin** [comment]
> is it possible using gen scope to define a reusable function that would be declared once in a global scope of a schema validator compilation instead of inline reinjection the same declaration clone per each keyword reference.

yes - please check the implementations of bundled keywords - there are quite a few cases where functions are referenced, e.g. “equal” - it’s possible both for compiling into closures (with ajv.compile) and for generating “standalone” validation code (in which case these functions should have require path as their property. Also this: https://ajv.js.org/codegen.html

> Maybe there's something it.root.gen available for this purpose?

It’s not needed - you can add to the shared scope via gen you have passed in the context - check out codegen.ts / scope.ts API and bundled keywords

> Do I see it correct as CodeKeywordDefinition is something the former inline keywords definition are evolved with time to?

the objective is the same - to generate code, but it’s not evolution. Old online was simply generating code as strings, without a simple way to modify shared scope, codegen generates lightweight syntax trees that are then optimised before rendering into strings. The latter is substantially slower (like 5x) but safer and produces smaller, more efficient code - so you have to make sure you compile once, use many times. Apparently some users were recompiling on every usage, so slower compilation affected runtime performance because of incorrect usage.

**Comment by artembeep** [comment]
@epoberezkin Yes awesome it really works! thanks.
Just to be sure could these thee be imported directly from 'ajv' somehow?
```js
import { useFunc } from 'ajv/dist/compile/util';
import standaloneCode from 'ajv/dist/standalone';
import { SchemaEnv } from 'ajv/dist/compile';
```
And I've confused a little with `useFunc` second argument notation. I've ended with some hack like this
```js
const myHelper = (a) =>  a + 1;
// ...

const ref = useFunc(gen, {code: String(myHelper)})
```
So maybe there's any other more appropriate wrapper to put a plain function to it?
